torchgeo/trainers/segmentation.py

```diff
@@ -111,8 +104,12 @@ def configure_losses(self) -> None:
                 ignore_index=ignore_value, weight=self.hparams["class_weights"]
             )
         elif loss == "jaccard":
+            class_set = set(range(self.hparams["num_classes"]))
+            if ignore_index is not None:
+                class_set.remove(ignore_index)
```

Comment:
@robmarkcole -- if the user sets `ignore_index` then it is removed from the `class_set` here and the remaining classes are  used in JaccardLoss.

Reply:
But `ignore_index` is a pixel value in the mask (e.g. 255) to ignore, unless I am mistaken?

Reply:
Correct, I've seen nodata as -9999, 0, 255 etc

Reply:
Just to be clear, my reasoning for opening this issue in the first place was that maybe you have a background class in the target masks that you don't care about explicitly learning, and so you ignore it in the loss even if you end up outputting it.

For instance, suppose I don't want to include the background class of LamdCoverAI in the loss. Then, I would just pass the corresponding index to `ignore_index`, and regardless of how many classes I chose to output, said class wouldn't be included in loss calculations...

Reply:
I don't actually know how that is handled by JaccardLoss or others. Let me check it out and come back.

Reply:
I'm back -- `JaccardLoss` assumes contiguous class values:

It sets `num_classes` to the size of the predicted mask's class dimension -- https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/losses/jaccard.py#L69

Then it converts `y_true` to a 1-hot encoding with this number of classes -- https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/losses/jaccard.py#L80

I confirmed this will throw an error if you have a mask with values like 255 but predictions with only 4 channels.

Reply:
Seems like ignore_index is the way to go so you would transform the integer mask to some values that you want to ignore in the loss. Too bad JaccardLoss doesn't have this as an input but I'm currently modifying it in torchseg.

This PR is probably okay as a workaround for now though

Reply:
@DimitrisMantas, yep totally understand. We're now thinking about the case where you have a mask where the values **aren't** {0, ..., C-1}. This happens often in practice (for example, NLCD land cover data have values 11, 12, 21, ... for whatever reason -- https://www.mrlc.gov/data/legends/national-land-cover-database-class-legend-and-description). This is probably out of scope for this particular PR as this case isn't support by smp's JaccardLoss anyway.

Reply:
If `ignore_index` is a large value like 255, on calling `class_set.remove(ignore_index)` a KeyError will be raised. Overall I think assuming continuous values will cause some issues and I think it would be better to wait for correct ignore_index handling in the loss. In my own code for now: 
```python
        if ignore_index:
            raise Exception("JaccardLoss does not observe ignore_index and one has been configured.")
        return smp.losses.JaccardLoss(mode="multiclass")
```

Reply:
We could use `discard` instead of `remove` to handle the case where `ignore_index` is not included in `range(self.hparams["num_classes"])`. But I agree that this does not the underlying issue of non-contiguous class labels...

Another way would be to go with the [`smp.FocalLoss`](https://github.com/qubvel/segmentation_models.pytorch/blob/6db76a1106426ac5b55f39fba68168f3bccae7f8/segmentation_models_pytorch/losses/focal.py#L78) approach:

```python
loss = 0

# Filter anchors with -1 label from loss computation
if self.ignore_index is not None:
	not_ignored = y_true != self.ignore_index

for cls in range(num_classes):
	cls_y_true = (y_true == cls).long()
	cls_y_pred = y_pred[:, cls, ...]

	if self.ignore_index is not None:
		cls_y_true = cls_y_true[not_ignored]
		cls_y_pred = cls_y_pred[not_ignored]

	loss += self.focal_loss_fn(cls_y_pred, cls_y_true)
```

However, there is a slight difference between this and the proposed approach. Namely, I believe `class_set` is effectively an index to the output channels, whereas `not_ignored` specifies certain target classes to ignore. Here is a relevant [test](https://gist.github.com/DimitrisMantas/658e16bf9c651588474550a1ad5d76ca).